labelme/shape.py

```diff
@@ -71,18 +77,65 @@ def shape_type(self, value):
         if value is None:
             value = 'polygon'
         if value not in ['polygon', 'rectangle', 'point',
-           'line', 'circle', 'linestrip']:
+           'line', 'circle', 'linestrip', 'curve']:
             raise ValueError('Unexpected shape_type: {}'.format(value))
         self._shape_type = value
 
     def close(self):
         self._closed = True
 
-    def addPoint(self, point):
-        if self.points and point == self.points[0]:
-            self.close()
+    def addSegment(self, seg_begin, seg_len):
+        if len(self.segments) == 0:
+            self.points_in_segments = 0
+            self.points_in_segments += seg_len
         else:
+            self.points_in_segments += (seg_len - 1)
+        new_segment = [seg_begin, seg_len]
+        self.segments.append(new_segment)
+
+    def addPointToSegment(self, point, seg_id):
+        self.segments[seg_id][1] += 1
+
+    def createSegment(self, point, degree_increment, new_p=0):
+        if point == self.points[-1]:
+            size = QUADRATIC_BEZIER_ORDER + degree_increment
+            seg_begin = len(self.points) - size
+            seg_len = size
+        else:
+            size = CUBIC_BEZIER_ORDER + degree_increment
+            self.insertPoint(max(len(self.points) - 1, 1), new_p)
+            seg_begin = len(self.points) - size
+            seg_len = size
             self.points.append(point)
+        self.addSegment(seg_begin, seg_len)
+
+    def addPoint(self, point, is_release=False, new_p=0):
+        if is_release and self.shape_type == 'curve':
+            if point == self.points[-1] and len(self.points) == 1:
+                self.points_in_segments += 1
+            elif len(self.segments) == 0 and len(self.points) == 1:
+                self.points.append(point)
+                self.points_in_segments += 1
+            else:
+                try:
+                    degree_increment = max(len(self.points) - self.points_in_segments - 1, 0)
+                    self.createSegment(point, degree_increment, new_p)
+                    if point == self.points[0]:
+                        self.points.pop(-1)
+                        self.close()
+                except NoPointError as e:
+                    assert False, "new_p is empty !!!"
```

Comment:
Asserts are debugging feature only, so throw an exception forward and just log an error message